proposals/0012-introduce-reactNativeMetadata.md

```diff
@@ -0,0 +1,329 @@
+
+### `align-deps` support
+
+This RFC originated from a [conversation between Meta and Microsoft](https://github.com/microsoft/rnx-kit/issues/1863) to use `align-deps` as key feature in the New Architecture rollout support, in helping developers understanding if a library they're using is New Architecture compatible or not. Very quickly it became clear that there is currently no straightforward way to know if a library indeed has that support.
+
+Therefore we believe `align-deps` can benefit from this information and be extended to offer New Architecture support information.
+
+### React Native CLI support
+
+Ideally the CLI (or any other build tool) should be able to consume the `reactNativeManifest` section and warn at build time if a configuration is invalid (either by having an invalid value or by having a mixture of invalid configurations).
+
+It's outside the scope of this RFC to delve into details of which tool will implement which logic and we defer this to a future discussion.
+
+### React Native Directory
+
+The React Native direcotry can easily consume the `reactNativeManifest` by:
+1. Query the NPM registry with [their API](https://github.com/npm/registry/blob/master/docs/REGISTRY-API.md) to obtain the repository information.
+2. Query the repository information to obtain the `packge.json` and read the `reactNativeManifest` section.
+
+## Future proof and extensibility
+
+We believe that the `reactNativeManifest` section should be extensible as we believe new capabilities and features can be added there.
+Specifically, we envision to evolve `reactNativeManifest` or `reactNativeManifest.capabilites` as follows:
+
+1. Each addition of a new capability [requires a new RFC](https://github.com/react-native-community/discussions-and-proposals#proposals) to be approved.
+2. Each RFC should define a new section in the `reactNativeManifest` with what's the intended use case.
+3. Each RFC should define what are the sensible defaults and should refrain from breaking changes.
+4. Each RFC should define the deprecation strategy for previous flags/features related to the same capability.
+
+## Complete Example
+
+Here we present a full example of how a `reactNativeManifest` section could look like for a library and an app.
+
+For an app the section will look as follows:
+    
+```json
+{
+  "name": "my-awesome-app",
+  "version": "1.2.3",
+  "reactNativeManifest": {
+    "version": "1.0",
+    "type": "app",
+    "capabilities": {
+      "hermes": {
+        "enabled": true
+      },
+      "newArch": {
+        "enabled": true
+      },
+      "codegenConfig": {
+        "name": "AppSpecs",
+        "type": "all",
+        "jsSrcsDir": ".",
+        "android": {
+          "javaPackageName": "com.facebook.fbreact.specs"
+        }
+      }
+    }
+  }
+}
+```
+
+For a library the section will look as follows:
+
+```json
+{
+  "name": "my-awesome-library",
+  "version": "1.2.3",
+  "reactNativeManifest": {
+    "version": "1.0",
```

Comment:
Something that came up in discussion is what is the correct way to version and show semantic changes in a native module.  Currently the docs for TurboModules imply versioning your native module using the package.json's version for[iOS](https://reactnative.dev/docs/the-new-architecture/pillars-turbomodules#ios-create-the-podspec-file) (I'm not seeing anything like this for Android).  There's some work taking place to surface incompatibility between native modules built (dev mode) into the native app and their JS counterparts bundled by Metro.  

Examples would be:
- forgetting to build in a native module and giving clearer error messages; or
- changing branch which includes a significant library change and metro rebuilding your JS but the developer forgetting to rebuild your native app.

## Questions:
1. Should native extensions be using the `package.json → reactNativeManifest → version` or some combination of both the `package.json → version` and the `reactNativeManifest → version`?
2. Would we update the docs, file tickets with significant [new architecture Native Modules implementations](https://reactnative.directory/?ios=true&android=true&isMaintained=true&isPopular=true&newArchitecture=true)

Reply:
See from the discussion with Nick that `reactNativeManifest → version` is a [schema version](https://github.com/react-native-community/discussions-and-proposals/pull/588#discussion_r1087184428).

Reply:
Yup also clarified here fe3e609f94490d9ac50213f5a16f71764c675d8a